File: node/src/resources/transfers.ts
import type { PaydClient } from "../client";
import type { TransferParams, TransferResponse } from "../types/transfers";
import { validateInternationalPhone } from "../validators/phone";
import { validatePositiveAmount } from "../validators/amount";
import { validateRequired } from "../validators/params";

/**
 * Payd-to-Payd transfer operations — send money between Payd accounts.
 */
export class Transfers {
  constructor(private readonly client: PaydClient) {}

  /**
   * Send money instantly between Payd accounts.
   *
   * Transfers are immediate — no webhook is needed (though one is sent if
   * a callback_url was originally set on the account).
   *
   * @example
   * ```typescript
   * const result = await payd.transfers.send({
   *   receiverUsername: "recipient_user",
   *   amount: 100,
   *   narration: "Lunch money",
   *   phoneNumber: "+254700000000",
   * });
   * console.log(result.transactionReference); // "9BD12041887.eS"
   * ```
   */
  async send(params: TransferParams): Promise<TransferResponse> {
    const walletType = params.walletType || this.client.defaults.walletType || "local";

    validateRequired(
      {
        receiverUsername: params.receiverUsername,
        narration: params.narration,
      },
      ["receiverUsername", "narration"],
    );

    validatePositiveAmount(params.amount);

    if (params.phoneNumber) {
      validateInternationalPhone(params.phoneNumber);
    }

    const body: Record<string, unknown> = {
      receiver_username: params.receiverUsername,
      amount: params.amount,
      narration: params.narration,
    };

    if (params.phoneNumber) {
      body.phone_number = params.phoneNumber;
    }
    if (params.currency) {
      body.currency = params.currency;
    }
    if (walletType !== "local") {
      body.wallet_type = walletType;
    }

    const data = await this.client.request({
      method: "POST",
      path: "/api/v2/remittance",
      body,
    });

    return {
      success: data.success as boolean,
      message: data.message as string,
      status: data.status as string,
      transactionReference:
        (data.transaction_reference as string) ||
        (data.correlator_id as string) ||
        "",
      trackingId: (data.trackingId as string) || "",
      reference: (data.reference as string) || "",
      result: data.result ?? null,
      _raw: data,
    };
  }
}

File: python/src/payd/resources/transfers.py
"""Payd-to-Payd transfer operations."""

from __future__ import annotations

from payd.validators import validate_international_phone, validate_positive_amount, validate_required


class Transfers:
    """Resource namespace for P2P transfers."""

    def __init__(self, client: "PaydClient") -> None:
        self._client = client

    def send(
        self,
        receiver_username: str,
        amount: int | float,
        narration: str,
        phone_number: str | None = None,
        currency: str | None = None,
        wallet_type: str | None = None,
    ) -> dict:
        """
        Send money from one Payd wallet to another.

        Optional values are included only when provided.
        """
        validate_required(
            {"receiver_username": receiver_username, "narration": narration},
            ["receiver_username", "narration"],
        )
        validate_positive_amount(amount)

        if phone_number:
            validate_international_phone(phone_number)

        resolved_wallet_type = wallet_type or self._client.defaults.get("wallet_type") or "local"

        body: dict[str, str | int | float] = {
            "receiver_username": receiver_username,
            "amount": amount,
            "narration": narration,
        }
        if phone_number:
            body["phone_number"] = phone_number
        if currency:
            body["currency"] = currency
        if resolved_wallet_type != "local":
            body["wallet_type"] = resolved_wallet_type

        return self._client.request(method="POST", path="/api/v2/p2p", body=body)
